### PR TITLE
Critdim

### PR DESCRIPTION
**Pull request by leahmariefernandez**
### Summary
setup project for the CRIT DIM

### Changelist 
- create CubeMX project + configure via referencing Altium (in directory firmware/quadruna/crit/cubemx/CRIT.ioc with set up RTOS tasks
- setup JSONCAN project
- set up CMake and integrate with the build scripts

### Testing Done
none

### Resolved Issues
none

### Checklist
*Please change `[ ]` to `[x]` when you are ready.*
- [x] I have read and followed the code conventions detailed in [README.md](../README.md) (*This will save time for both you and the reviewer!*).
- [x] If this pull request is longer then **500** lines, I have provided *explicit* justification in the summary above explaining why I *cannot* break this up into multiple pull requests (*Small PR's are faster and less painful for everyone involved!*).

### CODE REVIEW

File: firmware/quadruna/CRIT/CMakeLists.txt [hunk 1]
```diff
@@ -30,7 +33,7 @@ list(APPEND HW_SRCS
     "${SHARED_HW_INCLUDE_DIR}/hw_hardFaultHandler.c"
     "${SHARED_HW_INCLUDE_DIR}/hw_adcConversions.c"
     "${SHARED_HW_INCLUDE_DIR}/hw_can.c"
-    "${SHARED_HW_INCLUDE_DIR}/hw_bootup.c"
+    # "${SHARED_HW_INCLUDE_DIR}/hw_bootup.c"
```
Comment (Lucien950) [inline firmware/quadruna/CRIT/CMakeLists.txt:36]: please do include this file (I mention usage of this file in tasks.c)

File: firmware/quadruna/CRIT/src/tasks.c [hunk 2]
```diff
@@ -114,9 +304,66 @@ AdcChannel id_to_adc[] = {
...
+    [DIM_HEARTBEAT_BOARD]  = NULL,
+    [CRIT_HEARTBEAT_BOARD] = NULL
+};
+
 void tasks_preInit(void)
 {
-    // TODO: Setup bootloader.
+    // Setup bootloader.
+    // hw_bootup_enableInterruptsForApp();
 }
```
Comment (Lucien950) [inline firmware/quadruna/CRIT/src/tasks.c:367]: just uncomment this and add `#include "hw_bootup.h"` at the top.

File: firmware/quadruna/CRIT/src/app/app_globals.c [hunk 3]
```diff
@@ -0,0 +1,9 @@
+#include "app_globals.h"
+
+Globals        crit_globals;
+Globals *const globals = &crit_globals;
```
Comment (Lucien950) [inline firmware/quadruna/CRIT/src/app/app_globals.c:4]:
```suggestion
const Globals * globals = &crit_globals;
```

File: firmware/quadruna/CRIT/src/app/app_globals.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    const RgbLed *const    crit_status_led;
+    const RgbLed *const    rsm_status_led;
+} GlobalsConfig;
+
+typedef struct
+{
+    const GlobalsConfig *config;
+} Globals;
+
+extern Globals *const globals;
```
Comment (Lucien950) [inline firmware/quadruna/CRIT/src/app/app_globals.h:32]: you might have to fact check me on this, but i think we can `extern const crit_globals` (which can be renamed to just `globals`) while making `crit_globals` non-constant in the source file (.c file)

File: firmware/quadruna/CRIT/src/app/app_mainState.c [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    const bool bspd_fault_latched = app_canRx_BMS_BspdLatchedFault_get();
+    io_led_enable(globals->config->bspd_led, bspd_fault_latched);
+
+    // TODO: AMS fault
+    // const bool ams_fault_latched =
+    // io_led_enable(globals->config->ams_led, ams_fault_latched);
+
+    const bool contactors_open = app_canRx_BMS_AirNegative_get() == CONTACTOR_STATE_OPEN &&
+                                 app_canRx_BMS_AirPositive_get() == CONTACTOR_STATE_OPEN;
+    io_led_enable(globals->config->shdn_led, contactors_open);
```
Comment (Lucien950) [inline firmware/quadruna/CRIT/src/app/app_mainState.c:27]: fact check with Sumrath, but I believe shdn_led should be driven by the `SDHN_SEN` pin on the MCU.

File: firmware/quadruna/CRIT/src/app/app_mainState.c [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    };
+
+    CanAlertBoard alert_board_ids[NUM_BOARD_LEDS] = {
+        [BMS_LED] = BMS_ALERT_BOARD,
+        [FSM_LED] = FSM_ALERT_BOARD,
+        [VC_LED]  = VC_ALERT_BOARD,
+        //[AUX_LED] = AUX_ALERT_BOARD
+        [CRIT_LED] = CRIT_ALERT_BOARD,
+        //[RSM_LED] = PDM_ALERT_BOARD
+    };
```
Comment (Lucien950) [inline firmware/quadruna/CRIT/src/app/app_mainState.c:54]: Little datastructure nitpick, but I think this is hard to read. Try
```
struct LedCAN {
     RgbLed * led;
     CanAlertBoard jsoncan_board_enum;
}

const LedCAN boards[NUM_BOARD_LEDS] = {
    {
        .led = globals->config->bms_status_led
        .jsoncan_board_enum = BMS_ALERT_BOARD
    },
   ...
}
```
naming conventions you should improve on.

File: firmware/quadruna/CRIT/src/cubemx/Inc/FreeRTOSConfig.h [hunk 7]
```diff
@@ -175,6 +175,7 @@ standard names. */
 
 /* USER CODE BEGIN Defines */
 /* Section where parameter definitions can be added (for instance, to override default ones in FreeRTOS.h) */
+#include "SEGGER_SYSVIEW_FreeRTOS.h"
```
Comment (Lucien950) [inline firmware/quadruna/CRIT/src/cubemx/Inc/FreeRTOSConfig.h:178]: did you make this change? or was it autogen?

**Review by Lucien950 — CHANGES_REQUESTED** [review]
lots of good things. a few changes are requested. please make sure to do a basic flash validation on Saturday before we flash the boards together. also make sure to go through all the pins and make sure they are all used.

**Review by Lucien950 — CHANGES_REQUESTED** [review]
small watchdog thing

File: firmware/quadruna/CRIT/src/hw/hw_watchdogConfig.c [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+#include "io_canTx.h"
+#include "hw_utils.h"
+
+// TODO: Re-enable watchdog.
+// extern IWDG_HandleTypeDef hiwdg;
+
+void hw_watchdogConfig_refresh(void)
+{
+    // HAL_IWDG_Refresh(&hiwdg);
+}
```
[the remark this answers is not in the document]
Comment (Lucien950) [reply firmware/quadruna/CRIT/src/hw/hw_watchdogConfig.c:15]: btw if you don't reenable, hw watchdog will trip a lot

File: firmware/quadruna/CRIT/src/cubemx/Inc/FreeRTOSConfig.h (hunk 7 above)
Comment (leahmariefernandez) [reply firmware/quadruna/CRIT/src/cubemx/Inc/FreeRTOSConfig.h:178]: This was autogen I believe.

File: firmware/quadruna/CRIT/src/hw/hw_watchdogConfig.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (leahmariefernandez) [reply firmware/quadruna/CRIT/src/hw/hw_watchdogConfig.c:15]: Waiting until all board are finished to reenable.

File: firmware/quadruna/CRIT/src/cubemx/Inc/FreeRTOSConfig.h (hunk 7 above)
Comment (gtaharaedmonds) [reply firmware/quadruna/CRIT/src/cubemx/Inc/FreeRTOSConfig.h:178]: This is not autogen - it's a necessary change to get segger systemview working (our RTOS profiling tool)

File: firmware/quadruna/CRIT/src/app/app_globals.h (hunk 4 above)
Comment (leahmariefernandez) [reply firmware/quadruna/CRIT/src/app/app_globals.h:33]: Should be a pointer.

File: firmware/quadruna/CRIT/src/app/app_mainState.c (hunk 6 above)
Comment (leahmariefernandez) [reply firmware/quadruna/CRIT/src/app/app_mainState.c]: commit = "ledCAN datastructure"

**Review by Lucien950 — APPROVED** [review]

**Pull request merged**